### PR TITLE
Reading comprehension model for QUAC dataset

### PR DESCRIPTION
**Pull request by eunsol**
for EMNLP 18 paper

### CODE REVIEW

**Comment by matt-gardner** [comment]
Thanks @eunsol for submitting this!  @DeNeutoy, do you want to look at this one too, as you're also looking at @eunsol's other PR, or would you like me to do it?  I'm fine either way.

**Comment by DeNeutoy** [comment]
@matt-gardner you could take this one, if you don't mind.

File: allennlp/models/reading_comprehension/dialog_qa.py [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+from allennlp.modules.matrix_attention.linear_matrix_attention import LinearMatrixAttention
+from allennlp.modules.input_variational_dropout import InputVariationalDropout
+from allennlp.nn import InitializerApplicator, util
+from allennlp.training.metrics import Average, BooleanAccuracy, CategoricalAccuracy
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+@Model.register("dialog_qa")
+class DialogQA(Model):
```
Comment (matt-gardner) [inline allennlp/models/reading_comprehension/dialog_qa.py]: Minor nitpick: We prefer to use [Google's definition of camel case](https://github.com/allenai/allennlp/blob/master/STYLE.md#naming), so this should be `DialogQa`.

File: allennlp/models/reading_comprehension/dialog_qa.py [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+                 text_field_embedder: TextFieldEmbedder,
+                 phrase_layer: Seq2SeqEncoder,
+                 residual_encoder: Seq2SeqEncoder,
+                 span_start_encoder: Seq2SeqEncoder,
+                 span_end_encoder: Seq2SeqEncoder,
+                 initializer: InitializerApplicator,
+                 dropout: float = 0.2,
+                 num_context_answers: int = 0,
+                 mask_lstms: bool = True) -> None:
+        super(DialogQA, self).__init__(vocab)
```
Comment (matt-gardner) [inline allennlp/models/reading_comprehension/dialog_qa.py]: Minor python point: this should just be `super().__init__(vocab)`; the `DialogQA, self` bit is unnecessary.

File: allennlp/models/reading_comprehension/dialog_qa.py [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+        The encoder that we will use to incorporate span start predictions into the passage state
+        before predicting span end.
+    span_end_encoder : ``Seq2SeqEncoder``
+        The encoder that we will use to incorporate span end predictions into the passage state.
+    dropout : ``float``, optional (default=0.2)
+        If greater than 0, we will apply dropout with this probability after all encoders (pytorch
+        LSTMs do not apply dropout to their last layer).
+    num_context_answers : ``int``, optional (default=0)
+        If greater than 0, the model will consider previous question answering context.
+    mask_lstms : ``bool``, optional (default=True)
```
Comment (matt-gardner) [inline allennlp/models/reading_comprehension/dialog_qa.py]: Did you actually try this?  I saw some speed up with very early versions of allennlp, but @nelson-liu failed to reproduce the speedup with current code.  Just curious.

File: allennlp/models/reading_comprehension/dialog_qa.py [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+                 span_end_encoder: Seq2SeqEncoder,
+                 initializer: InitializerApplicator,
+                 dropout: float = 0.2,
+                 num_context_answers: int = 0,
+                 mask_lstms: bool = True) -> None:
+        super(DialogQA, self).__init__(vocab)
+        self._num_context_answers = num_context_answers
+        self._text_field_embedder = text_field_embedder
+        self._phrase_layer = phrase_layer
+        self._matrix_attention = LinearMatrixAttention(200, 200, 'x,y,x*y')
```
Comment (matt-gardner) [inline allennlp/models/reading_comprehension/dialog_qa.py]: Where does 200 come from here?  That's the expected output dimension of `phrase_layer`, right?  Instead of hard-coding this, you can do:

```python
encoding_dim = phrase_layer.get_output_dim()
self._matrix_attention = LinearMatrixAttention(encoding_dim, encoding_dim, 'x,y,x*y')
```
and similarly use `encoding_dim` below, everywhere you've hard-coded 200.

File: allennlp/models/reading_comprehension/dialog_qa.py [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+        self._merge_atten = TimeDistributed(torch.nn.Linear(200 * 4, 200))
+
+        self._residual_encoder = residual_encoder
+        self._prev_ans_marker = torch.nn.Embedding((num_context_answers * 4) + 1, 10)
```
Comment (matt-gardner) [inline allennlp/models/reading_comprehension/dialog_qa.py]: What's this 10 from?  Putting into a variable with a good name makes it easier to know what's going on with the code, and easier to modify later.

File: allennlp/models/reading_comprehension/dialog_qa.py [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+
+        if num_context_answers > 0:
+            self._question_num_marker = torch.nn.Embedding(12, 10 * num_context_answers)
```
Comment (matt-gardner) [inline allennlp/models/reading_comprehension/dialog_qa.py]: Same with the 10 and 12 here - is this 10 related to the 10 above or not?  A variable name would make that obvious.

File: allennlp/models/reading_comprehension/dialog_qa.py [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+
+        self._followup_lin = torch.nn.Linear(200, 3)
+        self._merge_self_atten = TimeDistributed(torch.nn.Linear(200 * 3, 200))
+
+        self._span_start_encoder = span_start_encoder
+        self._span_end_encoder = span_end_encoder
+
+        self._span_start_predictor = TimeDistributed(torch.nn.Linear(200, 1))
+        self._span_end_predictor = TimeDistributed(torch.nn.Linear(200, 1))
+        self._span_yesno_predictor = TimeDistributed(torch.nn.Linear(200, 3))
```
Comment (matt-gardner) [inline allennlp/models/reading_comprehension/dialog_qa.py]: This is for "yes", "no" and "no answer"?

Comment (eunsol) [reply allennlp/models/reading_comprehension/dialog_qa.py]: line 87 is to predict three-way classification, "yes", "no", "not a yes no question"

File: allennlp/models/reading_comprehension/dialog_qa.py (hunk 6 above)
Comment (eunsol) [reply allennlp/models/reading_comprehension/dialog_qa.py]: yes 10 is, i'll fix it. 12 is the max number of QA pairs in the dataset, and this is fixed..

File: allennlp/models/reading_comprehension/dialog_qa.py (hunk 3 above)
Comment (eunsol) [reply allennlp/models/reading_comprehension/dialog_qa.py]: no i haven't tried it, just copy-and-pasted them... ;P

File: allennlp/models/reading_comprehension/dialog_qa.py [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+        self._phrase_layer = phrase_layer
+
+        encoding_dim = phrase_layer.get_output_dim()
+        max_turn_length = 12
+
+        self._matrix_attention = LinearMatrixAttention(encoding_dim, encoding_dim, 'x,y,x*y')
+        self._merge_atten = TimeDistributed(torch.nn.Linear(encoding_dim * 4, encoding_dim))
+
+        self._residual_encoder = residual_encoder
+        self._prev_ans_marker = torch.nn.Embedding((num_context_answers * 4) + 1, marker_embedding_dim)
```
Comment (matt-gardner) [inline allennlp/models/reading_comprehension/dialog_qa.py]: I think this belongs in the `if` block below, because this is only actually used if `num_context_answers > 0`.

File: allennlp/models/reading_comprehension/dialog_qa.py [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+                 span_end_encoder: Seq2SeqEncoder,
+                 initializer: InitializerApplicator,
+                 dropout: float = 0.2,
+                 num_context_answers: int = 0,
+                 marker_embedding_dim: int = 10,
+                 mask_lstms: bool = True) -> None:
+        super().__init__(vocab)
+        self._num_context_answers = num_context_answers
+        self._text_field_embedder = text_field_embedder
+        self._phrase_layer = phrase_layer
```
Comment (matt-gardner) [inline allennlp/models/reading_comprehension/dialog_qa.py]: If you follow my recommendation to remove the shape mangling that combines the batch size and the number of questions, you will probably need to do `self._phrase_layer = TimeDistributed(phrase_layer)`, and on `self._matrix_attention`.

File: allennlp/models/reading_comprehension/dialog_qa.py [hunk 10]
```diff
@@ -0,0 +1,407 @@
...
+        # Shape: (batch_size * max_qa_count, passage_length, encoding_dim)
+        passage_question_vectors = util.weighted_sum(encoded_question, passage_question_attention)
+
+        # We replace masked values with something really negative here, so they don't affect the
+        # max below.
+        masked_similarity = util.replace_masked_values(passage_question_similarity,
+                                                       question_mask.unsqueeze(1),
+                                                       -1e7)
+
+        passage_mask = passage_mask.unsqueeze(1).repeat(1, max_qa_count, 1).view(batch_size * max_qa_count, -1)
```
Comment (matt-gardner) [inline allennlp/models/reading_comprehension/dialog_qa.py]: Instead of this line, I _think_ you can just do `util.last_dim_softmax(question_passage_similarity, passage_mask.unsqueeze(1))` below.

File: allennlp/models/reading_comprehension/dialog_qa.py [hunk 11]
```diff
@@ -0,0 +1,407 @@
...
+                                          passage_question_vectors,
+                                          repeated_encoded_passage * passage_question_vectors,
+                                          repeated_encoded_passage * tiled_question_passage_vector],
+                                         dim=-1)
+
+        final_merged_passage = F.relu(self._merge_atten(final_merged_passage))
+
+        merged_passage_for_residual = self._variational_dropout(final_merged_passage)
+        residual_layer = self._variational_dropout(self._residual_encoder(merged_passage_for_residual, passage_mask))
+        self_atten_matrix = self._self_atten(residual_layer, residual_layer)
```
Comment (matt-gardner) [inline allennlp/models/reading_comprehension/dialog_qa.py]: It'd be easier to read this code if `attention` weren't abbreviated.

File: allennlp/models/reading_comprehension/dialog_qa.py [hunk 12]
```diff
@@ -0,0 +1,407 @@
...
+
+        mask = passage_mask.resize(batch_size * max_qa_count,
+                                   passage_length,
+                                   1) * passage_mask.resize(batch_size * max_qa_count, 1, passage_length)
+
+        # torch.eye does not have a gpu implementation, so we are forced to use the cpu one and .cuda()
```
Comment (matt-gardner) [inline allennlp/models/reading_comprehension/dialog_qa.py]: `torch.eye` now has a GPU implementation.  So, to get the mask you want, you can just do this:

```python
passage_self_mask = torch.eye(passage_length, device=self_attention_matrix.device).unsqueeze(0).unsqueeze(0)
self_attention_mask = (1 - passage_self_mask) * passage_mask.unsqueeze(-1) * passage_mask.unsqueeze(-2)
```

Those two lines _should_ work to replace your current lines 215 through 225 (assuming you've undone the shape mangling, so you have `(batch_size, num_questions, ...)` shapes).

File: allennlp/models/reading_comprehension/dialog_qa.py [hunk 13]
```diff
@@ -0,0 +1,407 @@
...
+                                torch.stack([span_start, span_end], -1).view(batch_size * max_qa_count, 2),
+                                mask=qa_mask.unsqueeze(1).expand(-1, 2).long())
+            # add a select for the right span to compute loss
+            gold_span_end_loc = []
+            span_end = span_end.view(-1).squeeze().data.cpu().numpy()
+            for i in range(0, batch_size * max_qa_count):
+                gold_span_end_loc.append(max(span_end[i] * 3 + i * passage_length * 3, 0))
+                gold_span_end_loc.append(max(span_end[i] * 3 + i * passage_length * 3 + 1, 0))
+                gold_span_end_loc.append(max(span_end[i] * 3 + i * passage_length * 3 + 2, 0))
+            gold_span_end_loc = torch.LongTensor(gold_span_end_loc)
```
Comment (matt-gardner) [inline allennlp/models/reading_comprehension/dialog_qa.py]: Instead of this, you can do `gold_span_end_loc = span_end.new(gold_span_end_loc)`, which will create the tensor on the right device. Same with the predicted tensor below.

File: allennlp/models/reading_comprehension/dialog_qa.py [hunk 14]
```diff
@@ -0,0 +1,407 @@
...
+        span_followup_logits = span_followup_logits.data.cpu().numpy()
+        for b_i in range(batch_size):  # pylint: disable=invalid-name
+            for j in range(passage_length):
+                val1 = span_start_logits[b_i, span_start_argmax[b_i]]
+                if val1 < span_start_logits[b_i, j]:
+                    span_start_argmax[b_i] = j
+                    val1 = span_start_logits[b_i, j]
+                val2 = span_end_logits[b_i, j]
+                if val1 + val2 > max_span_log_prob[b_i]:
+                    if j - span_start_argmax[b_i] > 30:
```
Comment (matt-gardner) [inline allennlp/models/reading_comprehension/dialog_qa.py]: It's better to have magic numbers like this specified by a variable in `__init__`.  So you could have something like `self._max_span_length = 30`, and use `self._max_span_length` here.

File: allennlp/models/reading_comprehension/dialog_qa.py [hunk 15]
```diff
@@ -0,0 +1,407 @@
...
+                p3_answer_marker: torch.IntTensor = None,
+                yesno_list: torch.IntTensor = None,
+                followup_list: torch.IntTensor = None,
+                metadata: List[Dict[str, Any]] = None) -> Dict[str, torch.Tensor]:
+        model_in_cuda = span_start.is_cuda
+        qa_mask = (1 - torch.eq(followup_list, -1)).view(-1)  # not all dialog has the same number of QA pairs.
+        batch_size, max_qa_count, max_q_len, _ = question['token_characters'].size()
+        _, _, elmo_max_q_len, _ = question['elmo'].size()
+        question['token_characters'] = question['token_characters'].view(batch_size * max_qa_count, max_q_len, -1)
+        question['elmo'] = question['elmo'].view(batch_size * max_qa_count, elmo_max_q_len, -1)
```
[the remark this answers is not in the document]
Comment (eunsol) [reply allennlp/models/reading_comprehension/dialog_qa.py]: can text field embedder takes multi-dimensional (>3) tensor? if i’m remembering correctly, the text_field_embedder complained for having a multi dimensional dimensional tensor (question is [batch_size, max_qa_per_dialog, max_q_len, max_char_in_word])

[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/models/reading_comprehension/dialog_qa.py]: Yes, it should be able to, if you pass `num_wrapping_dims=1`.  If this doesn't work, let me know, because it's supposed to work.
